### ISSUE TITLE
ActionDispatch::Routing::RouteSet::Generator mutates request.path_parameters

### ISSUE THREAD

**Issue by bquorning** [rails/rails#24392]
I’m having some trouble upgrading to Rails 5.0.0.beta3, and it looks like #21057 is to blame.

In a controller, I can trigger an `ActionController::UrlGenerationError` exception by calling e.g. `root_path` right before calling `url_for`. In a `show` action, I would get something like this message: `No route matches {:action=>"index", :controller=>"foo", :id=>"42"}`. Notice how `url_for` thinks the current `show` action is an `index` action… This is because `ActionDispatch::Routing::RouteSet::Generator` is mutating `request.path_parameters`.

The commits 1993e2ccbd7c5651278ea30bdc9d8034f5197945 and 1a14074fb525cde4a439a1fb7515fe417c37ff96 remove the `dup` of the `recall` and `options` hashes, and changes a subsequent call to `@recall.delete`.

But there are still several ways to mutate the original objects. I’m seeing these calls:

```rb
@options.delete
@options[:action]= 
@options[:controller]=
@options[key]=
@recall[:action]=
```

My problems are fixed by changing only `@recall = recall` into `@recall = recall.dup`, but I would rather we fixed the mutation problems instead.

r? @schneems

**Comment by maclover7** [comment]
Can you create an executable test script using one of the templates [here](https://github.com/rails/rails/tree/master/guides/bug_report_templates), that would demonstrate the problem you are seeing?

**Comment by bquorning** [comment]
I've tried writing a test case, and failed. And will absolutely try again :-)

But just looking at the `Generator` code, it's pretty obviously buggy. Even the linked commits’ messages and that PRs comments mention the risk of introducing subtle bugs.

**Comment by bquorning** [comment]
Executable test script here: https://gist.github.com/bquorning/83a7689c50500500fd94773c14808b74

I’ve marked the 1 line you can remove to make the test pass.

**Comment by maclover7** [comment]
Bug script fails for me on master. Would you be willing to work on a pull request for this?

**Comment by bquorning** [comment]
The `options` attribute is probably ok not to dup, as @schneems wrote in 1a14074fb525cde4a439a1fb7515fe417c37ff96:

> This change removes one dup, since it's literally right after we just dup-d the hash to pass into this constructor.

For the `recall` attribute, I can see two ways of fixing it:

 1. Call `dup` in the initializer. Loses the performance benefits described in 1993e2ccbd7c5651278ea30bdc9d8034f5197945.
 2. Substitute `recall` for a `dup`e right before mutating it. Should keep most of those performance gains intact.

I’ll try going with 2).

**Comment by pixeltrix** [comment]
@bquorning this a duplicate of #23019 - I've tried doing a fix in #23103 but found some regressions in the changes that aren't tested in the current test suite. Will take another look at it this week.

**Comment by bquorning** [comment]
I *think* a fix might be as easy as calling `@recall = @recall.dup` right before mutating here:

```rb
def normalize_action!
   if @options[:action] == 'index'.freeze
     @recall[:action] = @options.delete(:action)
   end
 end
```

Test coverage is pretty sparse, though, so I'm not sure if this introduces new regressions.

**Comment by pixeltrix** [comment]
@bquorning yeah, I sure we could just `dup` something but I'd rather we fixed it properly :smile:

**Comment by bquorning** [comment]
Yeah, I agree. My observation is just that `@recall` has the same object_id as the controller's `request.path_parameters`.

**Comment by schneems** [comment]
Thanks for the report. I was OOO last week. I'll try to take a look soon. That hash dup was pretty expensive :(

**Comment by TikiTDO** [comment]
@schneems Are you sure the recall dup was expensive? I just tested the example script above with memory_profile, both with and without dup, and the difference was 300 bytes. Is there a test case where this recall hash is significantly bigger?

**Comment by bquorning** [comment]
I think the concern is that the `dup` happens for every URL generation on the page, and the dup'ed hashes are identical (except they're not, since we mutate them – I'm confused).

**Comment by TikiTDO** [comment]
I did some investigation on the code base, and here are my results. Please let me know if this makes sense:

 - Investigation
   - `ActionDispatch::Routing::RouteSet#url_for` calls into `ActionDispatch::Routing::RouteSet#generate` [here](https://github.com/TikiTDO/rails/blob/eba0041c2132e46bf9e040b31059e020e590d2ae/actionpack/lib/action_dispatch/routing/route_set.rb#L721)
   - `ActionDispatch::Routing::RouteSet#generate` creates an ininialized `ActionDispatch::Routing::RouteSet::Generator`, and calls `#generate` [here](https://github.com/TikiTDO/rails/blob/eba0041c2132e46bf9e040b31059e020e590d2ae/actionpack/lib/action_dispatch/routing/route_set.rb#L674)
     - This is the only place in the code base where `ActionDispatch::Routing::RouteSet::Generator` is used
   - `ActionDispatch::Routing::RouteSet::Generator#generate` calls `ActionDispatch::Journey::Formatter#generate` [here](https://github.com/TikiTDO/rails/blob/eba0041c2132e46bf9e040b31059e020e590d2ae/actionpack/lib/action_dispatch/routing/route_set.rb#L643)
     - No Problems: 
       - `ActionDispatch::Routing::RouteSet::Generator#generate` dose `constraints = path_parameters.merge(options)` which is equivalent to `constraints = recall.merge(options)` [here](https://github.com/TikiTDO/rails/blob/master/actionpack/lib/action_dispatch/journey/formatter.rb#L16)
       - `ActionDispatch::Routing::RouteSet::Generator#generate` calls `ActionDispatch::Routing::RouteSet::Generator#extract_parameterized_parts` which does `parameterized_parts = recall.merge(options)` [here](https://github.com/TikiTDO/rails/blob/master/actionpack/lib/action_dispatch/journey/formatter.rb#L55)
     - Possible Problems:
       - Problem 1: `ActionDispatch::Routing::RouteSet::Generator#extract_parameterized_parts` does `route.parts.reverse_each.drop_while { |part| !options.key?(part) || (options[part] || recall[part]).nil?}` [here](https://github.com/TikiTDO/rails/blob/master/actionpack/lib/action_dispatch/journey/formatter.rb#L58)
       - Problem 2: `ActionDispatch::Routing::RouteSet::Generator#generate` iterates does `params = options.dup.delete_if do |key, _|` [here](https://github.com/TikiTDO/rails/blob/master/actionpack/lib/action_dispatch/journey/formatter.rb#L29)
 - Issues
   - `ActionDispatch::Routing::RouteSet::Generator` writes to @recall in two places: `normalize_recall!` and `normalize_action!`
     - Root Cause 1: `normalize_recall!` does `@recall[:action] ||= 'index'`. 
       - This is rendered moot because the next method, normalize_options! defaults the options[:action] key to 'index'
     - Root Cause 2: `normalize_action!` does `@recall[:action] = @options.delete(:action) if @options[:action] == 'index'.freeze`
       - This is rendered moot, because the very next operation will do a recall.merge(options) anyway

 - Conclusion
   - With the current setup, during `ActionDispatch::Routing::RouteSet::Generator#generate` recall will always have an action parameter, while options will not.
   - Fix 1: For (Problem 1): Since option should never have an :action parameter, the block should always return true if `part == :action`
   - Fix 2: For (Problem 2): Since option should never have an :action parameter, the delete_if block can check if `part == :action`
   - Removing Root Cause 1 and 2, and adding Fix 1 and 2 should maintain identical behavior characteristics, while not needing the dup

**Comment by jeremy** [comment]
Fix in #23103
